### PR TITLE
Improve truncation error message

### LINKED ISSUE #256: Improve SqlBulkCopy truncation error message
### Is your feature request related to a problem? Please describe.
When you try to BulkCopy a too large string, you receive following error from [here](https://github.com/dotnet/SqlClient/blob/master/src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs#L1676):
```
System.InvalidOperationException: String or binary data would be truncated.
```
This message is "faked" by the managed code to look like the message sql server sends.
A year ago, this message is now available in an [improved form](https://blogs.msdn.microsoft.com/sql_server_team/string-or-binary-data-would-be-truncated-replacing-the-infamous-error-8152/), at least if the exception is actually thrown by MSSQL.

### Describe the solution you'd like
Please make similiar improvements (column and value affected) to the managed exception.

### Describe alternatives you've considered
If you can attach a debugger, you can enable first chance exception breaking for `InvalidOperationException`, then you'll see the string that was too long when you look in the locals window of the throwing stack frame. But this is only applicable when you can debug it.

### PR DESCRIPTION
**Pull request by Suchiman**
Brings the error message to parity with SQL Server 2016+ (trace flag) / 2019 (default).
fixes #256

### CODE REVIEW

**Comment by Suchiman** [comment]
Ouch :/ CI isn't public

**Comment by cheenamalhotra** [comment]
@Suchiman we'll help you out with logs till then! Public CI is WIP ✌️

File: src/Microsoft.Data.SqlClient/netcore/src/Resources/SR.Designer.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ namespace System {
     // class via a tool like ResGen or Visual Studio.
     // To add or remove a member, edit your .ResX file then rerun ResGen
     // with the /str option, or rebuild your VS project.
-    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "15.0.0.0")]
+    [global::System.CodeDom.Compiler.GeneratedCodeAttribute("System.Resources.Tools.StronglyTypedResourceBuilder", "16.0.0.0")]
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Resources/SR.Designer.cs]: Please revert this change, as we don't intend to enforce VS 2019

Comment (Suchiman) [reply src/Microsoft.Data.SqlClient/netcore/src/Resources/SR.Designer.cs]: I don't think this means that VS 2019 is required, this just indicates the tool that last generated this file, but will revert.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs [hunk 2]
```diff
@@ -1518,10 +1518,14 @@ private object ConvertValue(object value, _SqlMetaData metadata, bool isNull, re
                         value = SqlParameter.CoerceValue(value, mt, out coercedToDataFeed, out typeChanged, false);
                         if (!coercedToDataFeed)
                         {   // We do not need to test for TextDataFeed as it is only assigned to (N)VARCHAR(MAX)
-                            int len = ((isSqlType) && (!typeChanged)) ? ((SqlString)value).Value.Length : ((string)value).Length;
-                            if (len > length / 2)
+                            string str = ((isSqlType) && (!typeChanged)) ? ((SqlString)value).Value : ((string)value);
+                            if (str.Length > length / 2)
```
Comment (Wraith2) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: You might want to change this to include an IsNullOrEmpty check, there's nothing to guarantee that the str value is not null.

Comment (Suchiman) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: It was carelessly accessing `.Length` before as well, so either it was broken before as well or it's somehow safe

Comment (Suchiman) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: Actually, it is checked against here https://github.com/dotnet/SqlClient/blob/c8b60fc389ab01b7affeed673b634993d3179bfc/src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs#L1414-L1421

Comment (Wraith2) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: If clr null was passed in but the isNull parameter was false that code would return the null, and yes if it was doing that before I'd say it was careless. But since you're already improving it why not add the null check to improve this bit as well?

Comment (Suchiman) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: IMO this would be a bug, in particular since the SQL Server column could be `NOT NULL` so just letting `null` slip through would only help along to have the bug crawl deeper into the code before it blows up.

**Comment by Suchiman** [comment]
Just noticed that the netfx build is a distinct copy.
Unsure about Localisations other than `en` and `de`.
One could copy from the sql server translations

| message_id | language_id | text |
| ---------- | ----------- | ---- |
| 2628 | 1033 | String or binary data would be truncated in table '%.*1s', column '%.*ls'. Truncated value: '%.*ls'. |
| 2628 | 1031 | Zeichenfolgen- oder Binärdaten werden in Tabelle "%1!", Spalte "%2!" abgeschnitten. Abgeschnittener Wert: %3!. |
| 2628 | 1036 | Les données de chaîne ou binaires seraient tronquées dans la table '%1!', colonne '%2!'. Valeur tronquée : '%3!'. |
| 2628 | 1041 | テーブル '%1!'、列 '%2!' で文字列またはバイナリ データが切り捨てられます。切り捨てられた値: '%3!'。 |
| 2628 | 1030 | Strengdata eller binære data afkortes i tabel '%1!', kolonne '%2!'. Afkortet værdi: '%3!'. |
| 2628 | 3082 | Los datos binarios o de la cadena se truncan en la columna "%2!" de la tabla "%1!". Valor truncado: "%3!". |
| 2628 | 1040 | I dati stringa o binari verranno troncati nella tabella '%1!', colonna '%2!'. Valore troncato: (%3!). |
| 2628 | 1043 | Tekenreeks of binaire gegevens worden afgekapt in de tabel %1!, kolom %2!. Afgekapte waarde: %3!. |
| 2628 | 2070 | A cadeia de carateres ou dados binários seriam truncados na tabela "%1!", coluna "%2!". Valor truncado: "%3!". |
| 2628 | 1035 | Merkkijono- tai binaaritiedot katkaistaisiin taulukon %1! sarakkeessa %2!. Katkaistu arvo: %3!. |
| 2628 | 1053 | Sträng- eller binärdata skulle trunkeras i tabell %1!, kolumn %2!. Trunkerat värde: %3!. |
| 2628 | 1029 | Řetězec nebo binární data v tabulce %1!, sloupci %2! by byla zkrácena. Zkrácená hodnota: %3!. |
| 2628 | 1038 | A karakterlánc vagy a bináris adatok csonkolva lesznek a(z) %1! tábla %2! oszlopában. Csonkolt érték: %3!. |
| 2628 | 1045 | Ciąg lub dane binarne zostałyby obcięte w tabeli „%1!”, kolumna „%2!”. Obcięta wartość: „%3!”. |
| 2628 | 1032 | Τα δεδομένα συμβολοσειράς ή τα δυαδικά δεδομένα θα περικόπτονταν στον πίνακα "%1!", στήλη "%2!". Τιμή περικοπής: "%3!". |
| 2628 | 1049 | Произойдет усечение строковых или двоичных данных в столбце "%2!" таблицы "%1!". Усеченное значение: "%3!". |
| 2628 | 1055 | Dize veya ikili veri '%1!' tablosu, '%2!' sütununda kesilmiş. Kesilmiş değer: '%3!'. |
| 2628 | 1046 | Os dados de sequência ou binários estão truncados na tabela '%1!', coluna '%2!'. Valor truncado: '%3!'. |
| 2628 | 1028 | 資料表 '%1!'，資料行 '%2!' 中的字串或二進位資料將會截斷。截斷的值: '%3!'。 |
| 2628 | 1042 | 테이블 '%1!'의 열 '%2!'에 있는 문자열 또는 이진 데이터는 잘립니다. 잘린 값: '%3!'. |
| 2628 | 2052 | 字符串或二进制数据将在表“%1!”，列“%2!”中被截断。截断值:“%3!”。 |
| 2628 | 1044 | Streng eller binære data blir avkortet i tabellen %1!, kolonne %2!. Avkortede verdier: %3!. |

File: src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.de.resx [hunk 3]
```diff
@@ -2818,7 +2818,7 @@
     <value>Der angegebene ColumnName '{0}' entspricht keiner Spalte in der Datenquelle.</value>
   </data>
   <data name="SQL_BulkLoadStringTooLong" xml:space="preserve">
-    <value>String- oder binary-Daten würden abgeschnitten.</value>
+    <value>String- oder binary-Daten würden in Tabelle '{0}', Spalte '{1}' abgeschnitten. Abgeschnittener Wert: '{2}'.</value>
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netfx/src/Resources/Strings.de.resx]: Please revert localization file changes, these changes will be handled internally!

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs (hunk 2 above)
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: I would say it's not part of this PR, so rather let it be as it is.
L1414 `isNull` check should be revisited for this one if there's a possibility of sweeping in Null `value` with `isNull=false`.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs [hunk 4]
```diff
@@ -1518,10 +1518,14 @@ private object ConvertValue(object value, _SqlMetaData metadata, bool isNull, re
...
                             {
-                                throw SQL.BulkLoadStringTooLong();
+                                if (str.Length > 100)
+                                {
+                                    str = str.Remove(100);
+                                }
+                                throw SQL.BulkLoadStringTooLong(_destinationTableName, metadata.column, str);
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: if the above condition is not met, and `str` is not truncated, The `Truncated Value:` would show entire string, is that correct behavior since you're matching it with SQL Server?

Comment (Suchiman) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: Yes, SQL Server does not make a difference.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: Can you give me a query file that you used to test it out - just for confirming the behavior?

Comment (Suchiman) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: Check out the announcement blog posts, it demonstrates this behavior https://blogs.msdn.microsoft.com/sql_server_team/string-or-binary-data-would-be-truncated-replacing-the-infamous-error-8152/

Comment (Suchiman) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: Oh i see, you might be right, i've read it wrong it seems

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs]: Thanks for linking the blog post, I was wondering where to begin validation.
So that brings us 2 interesting points for this change:
- This error message is available from SQL Server 2019 (back ported to SQL 2017 CU12 and SQL 2016 SP2 CU but not enabled by default). For SqlClient users this is the only error message applicable to all servers, but I don't see any issue and a good opportunity to address this issue now!
- Magic number **_100_** seems like the max characters _(upon discovery, no official information regarding the same)_ that the error message would contain as "**truncated value**" that prints characters to be inserted in destination column.

My only worry was _[also comes from SQL Server's change, can't really do anything about it]_:
- Printing first **_100_** characters only doesn't solve the purpose of large strings getting truncated, users wouldn't know what's going on for strings longer than 100 characters.

**TODO**: Please add a comment in code for this number truncation to ensure we can address any complaints in future.

**Comment by Wraith2** [comment]
If always encrypted is used this may potentially leak an unencrypted input value of the field which exceeds the length into exception log files. To avoid that the value print would need to be optional and default to off.

**Comment by cheenamalhotra** [comment]
Good point @Wraith2 .

@Suchiman Could you verify if data at this stage is unencrypted/raw when AE is enabled and if yes, then we'll have to take a special handle to that scenario. We should respect client's encryption mode and should not change error message in this case.

From server side, string is Encrypted when thrown in error, so that turned acceptable. From our "mocked" error message, we need to be careful.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlUtil.cs [hunk 5]
```diff
@@ -822,9 +822,9 @@ internal static Exception BulkLoadNonMatchingColumnName(string columnName, Excep
         {
             return ADP.InvalidOperation(System.SRHelper.GetString(SR.SQL_BulkLoadNonMatchingColumnName, columnName), e);
         }
-        internal static Exception BulkLoadStringTooLong()
+        internal static Exception BulkLoadStringTooLong(string tableName, string columnName, string tuncatedValue)
         {
-            return ADP.InvalidOperation(System.SRHelper.GetString(SR.SQL_BulkLoadStringTooLong));
+            return ADP.InvalidOperation(System.SRHelper.GetString(SR.SQL_BulkLoadStringTooLong, tableName, columnName, tuncatedValue));
```
Comment (Wraith2) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlUtil.cs]: tuncated => truncated

**Comment by Wraith2** [comment]
> We should respect client's encryption mode and should not change error message in this case.

The change to the error message to contain the column and length is still very valuable but the truncated value would have to be omitted.

**Comment by Suchiman** [comment]
So in cases where encryption is enabled, should i add a second message without `Truncated value: ` or set `truncatedValue` to `<encrypted>`?

**Comment by Wraith2** [comment]
Setting it to `<encrypted>` sounds reasonable to me.
